### PR TITLE
fix: use standard API for Azure VMs in VMSS with flexible orchestration mode

### PR DESCRIPTION
**Pull request by pdecat**
If VMs are managed by VMSS in flexible orchestration mode, the standard VM API must be used: https://learn.microsoft.com/en-us/answers/questions/1418158/rest-api-request-returns-badrequest-for-vmss

Without this change, the following error occurs when trying to read `power_state` of such VMs:
```
# steampipe query 'select power_state from azure_compute_virtual_machine_scale_set_vm'                                                 11:24:09

Error: azure: compute.VirtualMachineScaleSetVMsClient#GetInstanceView: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InvalidParameter" Message="Virtual Machine Scale Set VM instanceId must be a number." Target="instanceId" (SQLSTATE HV000)

+-------------+
| power_state |
+-------------+
+-------------+
```

With this change, this returns the power state properly:

```
# steampipe query 'select power_state from azure_compute_virtual_machine_scale_set_vm'
+-------------+
| power_state |
+-------------+
| running     |
+-------------+

```

### CODE REVIEW

File: azure/table_azure_compute_virtual_machine.go [hunk 1]
```diff
@@ -47,6 +47,12 @@ func tableAzureComputeVirtualMachine(_ context.Context) *plugin.Table {
 				Description: "The friendly name that identifies the virtual machine.",
 				Type:        proto.ColumnType_STRING,
 			},
+			{
+				Name:        "virtual_machine_scale_set",
```
Comment (pdecat) [inline azure/table_azure_compute_virtual_machine.go:51]: This is necessary to distinguish standard VMs from VMs managed by VMSS in flexible orchestration mode.

File: azure/table_azure_compute_virtual_machine_scale_set_vm.go [hunk 2]
```diff
@@ -328,13 +328,24 @@ func getAzureComputeVirtualMachineScaleSetVmInstanceView(ctx context.Context, d
...
-	if err != nil {
-		return nil, err
-	}
+	if h.ParentItem.(compute.VirtualMachineScaleSet).OrchestrationMode == compute.Flexible {
+		client := compute.NewVirtualMachinesClientWithBaseURI(session.ResourceManagerEndpoint, subscriptionID)
+		client.Authorizer = session.Authorizer
 
-	return op, nil
+		op, err := client.InstanceView(ctx, resourceGroupName, *virtualMachine.Name)
+		if err != nil {
```
Comment (ParthaI) [inline azure/table_azure_compute_virtual_machine_scale_set_vm.go:337]: Please include the error log statement.

File: azure/table_azure_compute_virtual_machine_scale_set_vm.go [hunk 3]
```diff
@@ -328,13 +328,24 @@ func getAzureComputeVirtualMachineScaleSetVmInstanceView(ctx context.Context, d
...
+			return nil, err
+		}
+		return op, nil
+	} else {
+		client := compute.NewVirtualMachineScaleSetVMsClientWithBaseURI(session.ResourceManagerEndpoint, subscriptionID)
+		client.Authorizer = session.Authorizer
+
+		op, err := client.GetInstanceView(ctx, resourceGroupName, virtualMachine.ScaleSetName, instanceId)
+		if err != nil {
```
Comment (ParthaI) [inline azure/table_azure_compute_virtual_machine_scale_set_vm.go:346]: Please include the error log statement.

**Review by ParthaI — COMMENTED** [review]
Hi @pdecat,

Thank you so much for raising the PR! The changes make perfect sense to me.

I have a few suggestions:

- Could you please add a comment above the [getAzureComputeVirtualMachineScaleSetVmInstanceView](https://github.com/turbot/steampipe-plugin-azure/pull/829/files#diff-eba9cd0b524810c5ceb564e6bf40ed89d96c4be9bb8a04e992d96f7ebdb802f6R319) function explaining why these changes are needed? It would be helpful to include a link to the [official discussion](https://learn.microsoft.com/en-us/answers/questions/1418158/rest-api-request-returns-badrequest-for-vmss) regarding the error, for future reference.
- Could you also please add the query results in the PR description after fixing the issue?

Thank you again!

File: azure/table_azure_compute_virtual_machine_scale_set_vm.go (hunk 2 above)
Comment (pdecat) [reply azure/table_azure_compute_virtual_machine_scale_set_vm.go:342]: Done!

**Review by ParthaI — APPROVED** [review]

**Pull request merged**